### ISSUE TITLE
beforeEach, afterEach hooks not called if run second time after previous at least one test failed

### ISSUE THREAD

**Issue by romanmikhailov** [jquery/qunit#870]
I have next suite:
```javascript
(function(module, test)
{
    var beforeEachHook = function(assert)
    {
        console.log("beforeEach: " + assert.test.testName);
    };

    var afterEachHook = function(assert)
    {
        console.log("afterEach: " + assert.test.testName);
    };

    module("Suite with one failed test",
        {
            beforeEach: beforeEachHook,
            afterEach: afterEachHook
        }
    );

    test("First unit test", function(assert)
    {
        assert.ok(true);
    });

    test("Second unit test", function(assert)
    {
        assert.ok(false);
    });

    test("Third unit test", function(assert)
    {
        assert.ok(true);
    });

}(QUnit.module, QUnit.test));
```

When I run it first time I received in console next messages:
```
beforeEach: First unit test
afterEach: First unit test
beforeEach: Second unit test
afterEach: Second unit test
beforeEach: Third unit test
afterEach: Third unit test
```
Result of execution:
```
First unit test - okay
Second unit test - failed
Third unit test - okay
```

When I run it second time I received only next messages:
```
beforeEach: Second unit test
afterEach: Second unit test
beforeEach: First unit test
afterEach: First unit test
```
Result of execution:
```
First unit test - okay
Second unit test - failed
Third unit test - okay
```

As you can see in second run `before each` and `after each` of `Third unit test` is not called.

```javascript
QUnit.version = "1.19.0";
```

As I noticed `Test#hooks` not add `beforeEach`, `afterEach` hooks as it not pass next condition:
```javascript
		if ( this.module.testEnvironment &&
				QUnit.objectType( this.module.testEnvironment[ handler ] ) === "function" ) {
			hooks.push( this.queueHook( this.module.testEnvironment[ handler ], handler ) );
		}
```
```javascript
QUnit.objectType( this.module.testEnvironment[ handler ] )
```
equal to `"undefined"`

As it possibly cleared in ran `Test#before` (I notice that "execution" of test starts before all `Test#hook` executed) in next lines:
```javascript
		if ( this.module.testEnvironment ) {
			delete this.module.testEnvironment.beforeEach;
			delete this.module.testEnvironment.afterEach;
		}
```

**Comment by leobalter** [comment]
Confirmed: http://jsbin.com/yiqazobulo/edit?html,js,console,output

Thanks for this comprehensive bug report. Would you have time to write a fix for it?

**Comment by leobalter** [comment]
This regression came from fdf00a4. I'll see how can I avoid both problems.

**Comment by romanmikhailov** [comment]
@leobalter what about we change current behavior:
```javascript
if ( this.module.testEnvironment ) {
    delete this.module.testEnvironment.beforeEach;
    delete this.module.testEnvironment.afterEach;
}
this.testEnvironment = extend( {}, this.module.testEnvironment );
```
with next:
```javascript
this.testEnvironment = extend( {}, this.module.testEnvironment );
delete this.testEnvironment.beforeEach;
delete this.testEnvironment.afterEach;
```

**Comment by jzaefferer** [comment]
Will be addressed by #873
